src/test/scala/nl.knaw.dans.easy.properties/app/repository/sql/QueryGeneratorSpec.scala

```diff
@@ -377,6 +380,248 @@ class QueryGeneratorSpec extends TestSupportFixture with MockFactory {
+
+    val expectedQuery =
+      """SELECT DISTINCT depositorId
+        |FROM Deposit
+        |INNER JOIN (
+        |  SELECT SimpleProperties.depositId
+        |  FROM SimpleProperties
+        |  INNER JOIN (
+        |    SELECT depositId, max(timestamp) AS max_timestamp
+        |    FROM SimpleProperties
+        |    WHERE key = ?
+        |    GROUP BY depositId
+        |  ) AS SimplePropertiesWithMaxTimestamp
+        |  ON SimpleProperties.timestamp = SimplePropertiesWithMaxTimestamp.max_timestamp
+        |  WHERE value = ?
+        |) AS SimplePropertiesSearchResult
+        |ON Deposit.depositId = SimplePropertiesSearchResult.depositId;""".stripMargin
+    val expectedSize = List("ingest-step", IngestStepLabel.FEDORA.toString)
+
+    query should equal(expectedQuery)(after being whiteSpaceNormalised)
+    values should have size expectedSize.size
+    forEvery(values zip expectedSize) { case (value, expectedValue) =>
+      setStringMock(value, expectedValue)
+    }
+  }
+
+  it should "render a query that searches for depositors that deposited deposits that sometime had this certain ingest step" in {
+    val filter = DepositorIdFilters(ingestStepFilter = Some(DepositIngestStepFilter(IngestStepLabel.FEDORA, SeriesFilter.ALL)))
+    val (query, values) = QueryGenerator.searchDepositors(filter)
+
+    val expectedQuery =
+      """SELECT DISTINCT depositorId
+        |FROM Deposit
+        |INNER JOIN (
+        |  SELECT DISTINCT depositId
+        |  FROM SimpleProperties
+        |  WHERE key = ?
+        |  AND value = ?
+        |) AS SimplePropertiesSearchResult
+        |ON Deposit.depositId = SimplePropertiesSearchResult.depositId;""".stripMargin
+    val expectedSize = List("ingest-step", IngestStepLabel.FEDORA.toString)
+
+    query should equal(expectedQuery)(after being whiteSpaceNormalised)
+    values should have size expectedSize.size
+    forEvery(values zip expectedSize) { case (value, expectedValue) =>
+      setStringMock(value, expectedValue)
+    }
+  }
+
+  it should "render a query that searches for depositors that deposited deposits with a certain 'latest state' and that sometime had this certain ingest step" in {
+    val filter = DepositorIdFilters(
+      stateFilter = Some(DepositStateFilter(StateLabel.SUBMITTED, SeriesFilter.LATEST)),
+      ingestStepFilter = Some(DepositIngestStepFilter(IngestStepLabel.FEDORA, SeriesFilter.ALL)),
+    )
+    val (query, values) = QueryGenerator.searchDepositors(filter)
+
+    val expectedQuery =
+      """SELECT DISTINCT depositorId
+        |FROM Deposit
+        |INNER JOIN (
+        |  SELECT State.depositId
+        |  FROM State
+        |  INNER JOIN (
+        |    SELECT depositId, max(timestamp) AS max_timestamp
+        |    FROM State
+        |    GROUP BY depositId
+        |  ) AS StateWithMaxTimestamp
+        |  ON State.timestamp = StateWithMaxTimestamp.max_timestamp
+        |  WHERE label = ?
+        |) AS StateSearchResult
```

Comment:
I wonder we could use a derived table (maintained with a stored procedure?) or view with the last state of a deposit. The view would reduce complexity of queries, the derived table would increase maintenance performance but reduce query performance.

Reply:
CREATE VIEW LastState AS
SELECT State.stateId, State.depositId, State.label, State.description, State.timestamp FROM State
  INNER JOIN (
    SELECT depositId, max(timestamp) AS max_timestamp
    FROM State
    GROUP BY depositId
  ) AS StateWithMaxTimestamp
  ON State.timestamp = StateWithMaxTimestamp.max_timestamp

same for SimpleProperties

Reply:
Great idea! As discussed offline, we'll do this in another issue (https://drivenbydata.atlassian.net/browse/EASY-2415)